### PR TITLE
Write Branch Config Deltas to Generation Document

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch:
- Pulls some logic from _state/settings_ into a new _core/settings_ package.
- Charm config updated under a branch is written as a delta into the generation document.

## QA steps

There is currently no branch configuration retrieval to CLI; it is forthcoming. These tests rely on interrogation of the branch's _generation_ document directly. Note also that there are warnings when setting values to be the same as the those existing in master - this will be addressed in another future patch.

Setup:
- `export JUJU_DEV_FEATURE_FLAGS=generations`.
- Bootstrap.
- `juju deploy redis`.
- `juju branch test-branch`.
- `juju config redis databases=4 --branch master`.
- `juju config redis --reset databases --branch test-branch`. Document indicates deletion:
```
        "charm-config" : {
                "redis" : [
                        {
                                "type" : 2,
                                "key" : "databases",
                                "old" : NumberLong(4)
                        }
                ]
        },
```
- `juju config redis databases=4 --branch test-branch`. Document has no-op delta preserving the first value we observed for the setting:
```
        "charm-config" : {
                "redis" : [
                        {
                                "type" : 1,
                                "key" : "databases",
                                "old" : NumberLong(4),
                                "new" : NumberLong(4)
                        }
                ]
        },
```
- `juju config redis databases=8 --branch master`.
- `juju config redis databases=2 --branch test-branch`. Document indicates modification with original master value as old:
```
        "charm-config" : {
                "redis" : [
                        {
                                "type" : 1,
                                "key" : "databases",
                                "old" : NumberLong(4),
                                "new" : NumberLong(2)
                        }
                ]
        },
```
- `juju config redis password=new-pass --branch test-branch`. Document indicates last _databases_ entry, plus the addition:
```
        "charm-config" : {
                "redis" : [
                        {
                                "type" : 1,
                                "key" : "databases",
                                "old" : NumberLong(4),
                                "new" : NumberLong(2)
                        },
                        {
                                "type" : 0,
                                "key" : "password",
                                "new" : "new-pass"
                        }
                ]
        },
```

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

File: core/settings/settings.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+// ItemChanges contains a slice of item changes in a config node.
+// It implements the sort interface to sort the items changes by key.
+type ItemChanges []ItemChange
+
+// ApplyDeltaSource turns this second-order delta into a first-older delta
+// by replacing the OldValue for any key present in the input ItemChange
+// collection.
+func (c ItemChanges) ApplyDeltaSource(oldChanges ItemChanges) error {
```
Comment (manadart) [inline core/settings/settings.go]: Not completely sold on this name; welcome suggestions.

File: state/application_test.go [hunk 2]
```diff
@@ -254,6 +254,8 @@ func (s *ApplicationSuite) TestSetCharmCharmSettings(c *gc.C) {
 }
 
 func (s *ApplicationSuite) TestGenerationCharmCharmSettings(c *gc.C) {
+	c.Skip("To be rewritten when branch-based configuration reads are implemented.")
```
Comment (manadart) [inline state/application_test.go]: I will reinstate this as a set/get test when reading branch-based charm configuration is implemented in a following patch.

**Comment by manadart** [comment]
[Store Branch Configuration Deltas](https://trello.com/c/hTo8wqif/1250-store-branch-configuration-deltas)

File: state/settings_test.go [hunk 3]
```diff
@@ -594,3 +568,38 @@ func (s *SettingsSuite) TestUpdatingInterfaceSliceValue(c *gc.C) {
...
+	s2.applyChanges(coresettings.ItemChanges{
+		coresettings.MakeModification("foo", "no-matter", "new-bar"),
+		coresettings.MakeModification("make", "no-matter", "new"),
+		coresettings.MakeDeletion("alpha", "no-matter"),
+		coresettings.MakeDeletion("what", "the"),
+		coresettings.MakeAddition("new", "noob"),
+	})
+
+	// Updating one not present = addition, deleting one not present = no-op.
+	exp := coresettings.ItemChanges{
```
Comment (howbazaar) [inline state/settings_test.go]: Where did "number" go?

File: state/modelgeneration.go [hunk 4]
```diff
@@ -43,6 +45,11 @@ type generationDoc struct {
 	// generation, but no units currently set to be in it.
 	AssignedUnits map[string][]string `bson:"assigned-units"`
 
+	// Config is all changes made to charm configuration under this branch.
+	Config model.BranchCharmConfig `bson:"charm-config"`
```
[the remark this answers is not in the document]
Comment (manadart) [reply state/modelgeneration.go]: Fair enough. I thought there was a precedent for this given a couple of struct types with BSON tags in _core/devices_, but now I see that they don't appear to be used inside the state package.

File: state/settings_test.go (hunk 3 above)
Comment (manadart) [reply state/settings_test.go]: Number is not affected by the application of the change-set, so it is not present in a delta; I.e. it remains as is.

File: core/settings/settings.go (hunk 1 above)
Comment (manadart) [reply core/settings/settings.go]: The logic actually needs to change for at lease one case currently not correct:
- A branch change is an _addition_.
- The master settings gets a value added for the same key.
- A _modification_ of the same key on the branch turns it into an update.

So for these, we need to preserve changes that were _additions_. This means that we are also changing the types as well as the old values.

Any suggestions with this in mind?

File: state/settings.go [hunk 5]
```diff
@@ -146,57 +119,74 @@ func (s *Settings) Delete(key string) {
...
+	var changes []settings.ItemChange
+
 	for key := range cacheKeys(s.disk, s.core) {
 		old, onDisk := s.disk[key]
 		live, inCore := s.core[key]
 		if reflect.DeepEqual(live, old) {
 			continue
 		}
-		var change ItemChange
-		escapedKey := utils.EscapeKey(key)
```
[the remark this answers is not in the document]
Comment (manadart) [reply state/settings.go]: It is done in `settingsUpdateOps` above this one. I have modified the new test to include keys that need escape/unescape handling.

File: state/modelgeneration.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (manadart) [reply state/modelgeneration.go]: I made a state-based type to use for persistence. I needed to do key escape/unescape there anyway.

File: state/modelgeneration.go [hunk 6]
```diff
@@ -318,7 +344,7 @@ func (g *Generation) UpdateCharmConfig(appName string, master *Settings, validCh
 					{{"txn-revno", g.doc.TxnRevno}},
 				}}},
 				Update: bson.D{
-					{"$set", bson.D{{"charm-config", g.doc.Config}}},
+					{"$set", bson.D{{"charm-config." + appName, makeItemChanges(newDelta)}}},
```
Comment (howbazaar) [inline state/modelgeneration.go]: I'm assuming here that if the appName didn't have any value before, this just does the right thing.

File: state/modelgeneration_test.go [hunk 7]
```diff
@@ -208,9 +210,60 @@ func (s *generationSuite) TestCommitNoChangesEffectivelyAborted(c *gc.C) {
 
 // TODO (manadart 2019-03-21): Tests for abort.
 
+func (s *generationSuite) TestBranchCharmConfigDeltas(c *gc.C) {
+	gen := s.setupAssignAllUnits(c)
+	c.Assert(gen.Config(), gc.HasLen, 0)
+
+	current := state.GetPopulatedSettings(map[string]interface{}{
```
Comment (howbazaar) [inline state/modelgeneration_test.go]: Worth adding a test for the escaping/unescaping? Just having a key with a '.' or '$' would be good.

File: state/settings_test.go [hunk 8]
```diff
@@ -585,18 +585,18 @@ func (s *SettingsSuite) TestApplyAndRetrieveChanges(c *gc.C) {
 	// Add, update, update one not present, delete, delete one not present,
 	// leave one alone.
 	s2.applyChanges(coresettings.ItemChanges{
-		coresettings.MakeModification("foo", "no-matter", "new-bar"),
+		coresettings.MakeModification("foo.dot", "no-matter", "new-bar"),
```
Comment (howbazaar) [inline state/settings_test.go]: Never mind, I see what I wanted to see here.

File: core/settings/settings.go [hunk 9]
```diff
@@ -84,21 +94,50 @@ func MakeDeletion(key string, oldVal interface{}) ItemChange {
 // It implements the sort interface to sort the items changes by key.
 type ItemChanges []ItemChange
 
-// ApplyDeltaSource turns this second-order delta into a first-older delta
-// by replacing the OldValue for any key present in the input ItemChange
-// collection.
-func (c ItemChanges) ApplyDeltaSource(oldChanges ItemChanges) error {
+// ApplyDeltaSource uses this second-order delta generate a first-older delta
+// by comparing with the incoming changes collection.
```
[the remark this answers is not in the document]
Comment (manadart) [reply core/settings/settings.go]: Added more detail regarding the requirement it addresses.

File: state/modelgeneration.go (hunk 6 above)
Comment (manadart) [reply state/modelgeneration.go]: Yes, works fine from a standing start.

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
Good ol' `WARNING juju.testing failed to start mongo: mongod failed to listen on port 40352`.

$$merge$$

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
@jameinel Looks like the pending comments were lost...

**Pull request merged**